### PR TITLE
Mark other_selected if Other selected

### LINKED ISSUE #2846: [Data collection] "Other" multiple choice option with no text not saved
The UI appears to be removing the value, resulting in https://github.com/google/ground-platform/issues/2099 and https://github.com/google/ground-platform/issues/2098.

We may need to add a bool field to `MultipleChoiceResponses` to indicate the "other" option was selected; we can't rely on `other_text` being set since proto3 does not have a separate "unset" state for optional fields, so an empty string is considered the same as an unset field.

@kenstershiro  @anandwana001 FYI

### PR DESCRIPTION
**Pull request by anandwana001**
Fixes #2846

Added proto in this PR
https://github.com/google/ground-platform/pull/2107

Screenshot 2024-11-27 at 11 10 08 AM

@shobhitagarwal1612  PTAL?

### CODE REVIEW

**Comment by anandwana001** [comment]
I don't think we need the other variable as mentioned here
https://github.com/google/ground-android/issues/2846#issuecomment-2498445587


Please let me know if I am missing anything here.

File: ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt [hunk 1]
```diff
@@ -139,6 +139,7 @@ private fun toTaskData(id: String, newTaskData: TaskData) = taskData {
     is MultipleChoiceTaskData -> multipleChoiceResponses = multipleChoiceResponses {
         newTaskData.getSelectedOptionsIdsExceptOther().forEach { selectedOptionIds.add(it) }
         if (newTaskData.hasOtherText()) {
+          otherSelected = true
           otherText = newTaskData.getOtherText()
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt:143]: This won't work if the text isn't set.

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt [hunk 2]
```diff
@@ -95,7 +95,13 @@ class MultipleChoiceTaskViewModel @Inject constructor() : AbstractTaskViewModel(
...
         selectedIds.map {
           // if the other option is selected grab the associated text save instead of otherId
-          if (it == OTHER_ID) OTHER_PREFIX + otherText.trim() + OTHER_SUFFIX else it
+          if (it == OTHER_ID) {
+            OTHER_PREFIX +
+              (if (otherText.isNotBlank()) otherText.trim() else "Other") +
+              OTHER_SUFFIX
+          } else {
+            it
+          }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:104]: Can we revert this?

File: ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt [hunk 3]
```diff
@@ -139,6 +139,7 @@ private fun toTaskData(id: String, newTaskData: TaskData) = taskData {
     is MultipleChoiceTaskData -> multipleChoiceResponses = multipleChoiceResponses {
         newTaskData.getSelectedOptionsIdsExceptOther().forEach { selectedOptionIds.add(it) }
         if (newTaskData.hasOtherText()) {
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt:141]: Can we rename `hasOtherText()` to `isOtherSelected()` for correctness?

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt (hunk 2 above)
Comment (anandwana001) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:104]: but without this the value will always empty, we have to show "Other" text, right?
Or it will be taken care by backend?

File: ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt (hunk 3 above)
Comment (anandwana001) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt:141]: not a part of this PR, but let me update at all places

File: ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt (hunk 2 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt:104]: We don't need a default value for the other text.

Comment (anandwana001) [reply ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt]: done

File: ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt (hunk 3 above)
Comment (anandwana001) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firebase/protobuf/ModelToProtoExt.kt]: done

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]
Please update unit tests and PR description as well

**Comment by anandwana001** [comment]
> Please update unit tests and PR description as well

Waiting for web pr https://github.com/google/ground-platform/pull/2107 to get in so for the proto variable

**Review by gino-m — APPROVED** [review]
🔢 👍

**Comment by shobhitagarwal1612** [comment]
Please add some test coverage to `SubmissionMutationConverterTest`

File: ground/src/test/java/com/google/android/ground/persistence/remote/firebase/schema/SubmissionMutationConverterTest.kt [hunk 4]
```diff
@@ -234,7 +235,10 @@ class SubmissionMutationConverterTest {
       ),
       mapOf(
         MULTIPLE_CHOICE_RESPONSES_FIELD_NUMBER.toString() to
-          mapOf(OTHER_TEXT_FIELD_NUMBER.toString() to "other value"),
+          mapOf(
+            OTHER_TEXT_FIELD_NUMBER.toString() to "other value",
+            OTHER_SELECTED_FIELD_NUMBER.toString() to true,
+          ),
         TASK_ID_FIELD_NUMBER.toString() to "single_choice_task_other",
```
Comment (shobhitagarwal1612) [inline ground/src/test/java/com/google/android/ground/persistence/remote/firebase/schema/SubmissionMutationConverterTest.kt:242]: Can we add one more test case where the other text is empty?

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]

File: ground/src/test/java/com/google/android/ground/persistence/remote/firebase/schema/SubmissionMutationConverterTest.kt (hunk 4 above)
Comment (anandwana001) [reply ground/src/test/java/com/google/android/ground/persistence/remote/firebase/schema/SubmissionMutationConverterTest.kt:259]: Done

**Review by shobhitagarwal1612 — APPROVED** [review]
